File: src/test/java/no/nav/syfo/azuread/AzureAdTokenConsumerTest.java
package no.nav.syfo.azuread;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import no.nav.syfo.LocalApplication;
import org.junit.*;
import org.junit.runner.RunWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.web.client.RestTemplate;

import javax.inject.Inject;
import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.client.ExpectedCount.once;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

@RunWith(SpringRunner.class)
@SpringBootTest(classes = LocalApplication.class)
@DirtiesContext
public class AzureAdTokenConsumerTest {
    @Inject
    private RestTemplate restTemplateMedProxy;

    private AzureAdTokenConsumer azureAdTokenConsumer;

    private MockRestServiceServer mockRestServiceServer;

    private final String TOKEN_URL = "https://url.nav.no";

    @Before
    public void setup() {
        this.mockRestServiceServer = MockRestServiceServer.bindTo(restTemplateMedProxy).build();
        azureAdTokenConsumer = new AzureAdTokenConsumer(restTemplateMedProxy, TOKEN_URL, "clientId", "clientSecret");
    }

    @After
    public void tearDown() {
        mockRestServiceServer.verify();
    }

    @Test
    public void henterTokenFraAzureHvisTokenMangler() {
        Instant expiresOn = Instant.now().plusSeconds(300L);
        String tokenGyldig = "token";
        String responseBody = azureAdResponseAsJsonString(expiresOn, "ressursID", tokenGyldig);
        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBody, MediaType.APPLICATION_JSON));

        final String token = azureAdTokenConsumer.getAccessToken("test");

        assertThat(token).isEqualTo(tokenGyldig);
    }

    @Test
    public void henterTokenFraAzureHvisTokenErUtlopt() {
        Instant expiresOnUtlopt = Instant.now().minusSeconds(30L);
        String tokenUtlopt = "token_utlopt";
        String responseBodyUtlopt = azureAdResponseAsJsonString(expiresOnUtlopt, "ressursID", tokenUtlopt);

        Instant expiresOnGyldig = Instant.now().plusSeconds(300L);
        String tokenGyldig = "token_gyldig";
        String responseBodyGyldig = azureAdResponseAsJsonString(expiresOnGyldig, "ressursID", tokenGyldig);

        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBodyUtlopt, MediaType.APPLICATION_JSON));
        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBodyGyldig, MediaType.APPLICATION_JSON));

        azureAdTokenConsumer.getAccessToken("test");
        final String token = azureAdTokenConsumer.getAccessToken("test");

        assertThat(token).isEqualTo(tokenGyldig);
    }

    @Test
    public void brukerEksisterendeTokenHvisGyldig() {
        Instant expiresOnGyldig = Instant.now().plusSeconds(300L);
        String tokenGyldig = "token_gyldig";
        String responseBodyGyldig = azureAdResponseAsJsonString(expiresOnGyldig, "ressursID", tokenGyldig);

        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBodyGyldig, MediaType.APPLICATION_JSON));

        azureAdTokenConsumer.getAccessToken("test");
        final String token = azureAdTokenConsumer.getAccessToken("test");

        assertThat(token).isEqualTo(tokenGyldig);
    }

    @Test
    public void henterTokenFraAzureHvisTokenForResourceMangler() {
        Instant expiresOn1 = Instant.now().minusSeconds(300L);
        String tokenForResource1 = "token_1";
        String resource1 = "resource_1";
        String responseBody1 = azureAdResponseAsJsonString(expiresOn1, resource1, tokenForResource1);

        Instant expiresOn2 = Instant.now().plusSeconds(300L);
        String tokenForResource2 = "token_2";
        String resource2 = "resource_2";
        String responseBody2 = azureAdResponseAsJsonString(expiresOn2, resource2, tokenForResource2);

        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBody1, MediaType.APPLICATION_JSON));
        mockRestServiceServer.expect(once(), requestTo(TOKEN_URL)).andRespond(withSuccess(responseBody2, MediaType.APPLICATION_JSON));

        final String token1 = azureAdTokenConsumer.getAccessToken(resource1);
        final String token2 = azureAdTokenConsumer.getAccessToken(resource2);
        final String token3 = azureAdTokenConsumer.getAccessToken(resource2);

        assertThat(token1).isEqualTo(token1);
        assertThat(token2).isEqualTo(token2);
        assertThat(token3).isEqualTo(token2);
    }

    private String azureAdResponseAsJsonString(Instant expiresOn, String resource, String token) {
        ObjectMapper objectMapper = new ObjectMapper();
        JavaTimeModule module = new JavaTimeModule();
        objectMapper.registerModule(module);

        AzureAdResponse azureAdResponse = new AzureAdResponse()
                .access_token(token)
                .token_type("Bearer")
                .expires_in("3600")
                .ext_expires_in("3600")
                .expires_on(expiresOn)
                .not_before(Long.toString(expiresOn.getEpochSecond()))
                .resource(resource);

        try {
            return objectMapper.writeValueAsString(azureAdResponse);
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/test/kotlin/no/nav/syfo/azuread/AzureAdTokenConsumerTest.kt
package no.nav.syfo.azuread

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import no.nav.syfo.LocalApplication
import org.assertj.core.api.Assertions
import org.junit.*
import org.junit.runner.RunWith
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.client.ExpectedCount
import org.springframework.test.web.client.MockRestServiceServer
import org.springframework.test.web.client.match.MockRestRequestMatchers
import org.springframework.test.web.client.response.MockRestResponseCreators
import org.springframework.web.client.RestTemplate
import java.time.Instant
import javax.inject.Inject

@RunWith(SpringRunner::class)
@SpringBootTest(classes = [LocalApplication::class])
@DirtiesContext
class AzureAdTokenConsumerTest {
    @Inject
    private lateinit var restTemplateMedProxy: RestTemplate
    private lateinit var azureAdTokenConsumer: AzureAdTokenConsumer
    private lateinit var mockRestServiceServer: MockRestServiceServer
    private val TOKEN_URL = "https://url.nav.no"

    @Before
    fun setup() {
        mockRestServiceServer = MockRestServiceServer.bindTo(restTemplateMedProxy).build()
        azureAdTokenConsumer = AzureAdTokenConsumer(restTemplateMedProxy, TOKEN_URL, "clientId", "clientSecret")
    }

    @After
    fun tearDown() {
        mockRestServiceServer.verify()
    }

    @Test
    fun henterTokenFraAzureHvisTokenMangler() {
        val expiresOn = Instant.now().plusSeconds(300L)
        val tokenGyldig = "token"
        val responseBody = azureAdResponseAsJsonString(expiresOn, "ressursID", tokenGyldig)
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBody, MediaType.APPLICATION_JSON))
        val token = azureAdTokenConsumer.getAccessToken("test")
        Assertions.assertThat(token).isEqualTo(tokenGyldig)
    }

    @Test
    fun henterTokenFraAzureHvisTokenErUtlopt() {
        val expiresOnUtlopt = Instant.now().minusSeconds(30L)
        val tokenUtlopt = "token_utlopt"
        val responseBodyUtlopt = azureAdResponseAsJsonString(expiresOnUtlopt, "ressursID", tokenUtlopt)
        val expiresOnGyldig = Instant.now().plusSeconds(300L)
        val tokenGyldig = "token_gyldig"
        val responseBodyGyldig = azureAdResponseAsJsonString(expiresOnGyldig, "ressursID", tokenGyldig)
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBodyUtlopt, MediaType.APPLICATION_JSON))
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBodyGyldig, MediaType.APPLICATION_JSON))
        azureAdTokenConsumer.getAccessToken("test")
        val token = azureAdTokenConsumer.getAccessToken("test")
        Assertions.assertThat(token).isEqualTo(tokenGyldig)
    }

    @Test
    fun brukerEksisterendeTokenHvisGyldig() {
        val expiresOnGyldig = Instant.now().plusSeconds(300L)
        val tokenGyldig = "token_gyldig"
        val responseBodyGyldig = azureAdResponseAsJsonString(expiresOnGyldig, "ressursID", tokenGyldig)
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBodyGyldig, MediaType.APPLICATION_JSON))
        azureAdTokenConsumer.getAccessToken("test")
        val token = azureAdTokenConsumer.getAccessToken("test")
        Assertions.assertThat(token).isEqualTo(tokenGyldig)
    }

    @Test
    fun henterTokenFraAzureHvisTokenForResourceMangler() {
        val expiresOn1 = Instant.now().minusSeconds(300L)
        val tokenForResource1 = "token_1"
        val resource1 = "resource_1"
        val responseBody1 = azureAdResponseAsJsonString(expiresOn1, resource1, tokenForResource1)
        val expiresOn2 = Instant.now().plusSeconds(300L)
        val tokenForResource2 = "token_2"
        val resource2 = "resource_2"
        val responseBody2 = azureAdResponseAsJsonString(expiresOn2, resource2, tokenForResource2)
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBody1, MediaType.APPLICATION_JSON))
        mockRestServiceServer.expect(ExpectedCount.once(), MockRestRequestMatchers.requestTo(TOKEN_URL)).andRespond(MockRestResponseCreators.withSuccess(responseBody2, MediaType.APPLICATION_JSON))
        val token1 = azureAdTokenConsumer.getAccessToken(resource1)
        val token2 = azureAdTokenConsumer.getAccessToken(resource2)
        val token3 = azureAdTokenConsumer.getAccessToken(resource2)
        Assertions.assertThat(token1).isEqualTo(token1)
        Assertions.assertThat(token2).isEqualTo(token2)
        Assertions.assertThat(token3).isEqualTo(token2)
    }

    private fun azureAdResponseAsJsonString(expiresOn: Instant, resource: String, token: String): String {
        val objectMapper = ObjectMapper()
        val module = JavaTimeModule()
        objectMapper.registerModule(module)
        val azureAdResponse = AzureAdResponse()
            .access_token(token)
            .token_type("Bearer")
            .expires_in("3600")
            .ext_expires_in("3600")
            .expires_on(expiresOn)
            .not_before(java.lang.Long.toString(expiresOn.epochSecond))
            .resource(resource)
        return try {
            objectMapper.writeValueAsString(azureAdResponse)
        } catch (e: JsonProcessingException) {
            throw RuntimeException(e)
        }
    }
}
